mypy/checker.py

```diff
@@ -2358,26 +2388,69 @@ def find_isinstance_check(node: Node,
+        the proposed type, if the expression can be the proposed type.  The
+        second element is a map from the expression to the type it would hold
+        if it was not the proposed type, if any."""
+        if proposed_type:
+            if current_type:
+                if is_proper_subtype(current_type, proposed_type):
+                    return {expr: proposed_type}, None
+                elif not is_overlapping_types(current_type, proposed_type):
+                    return None, {expr: current_type}
+                else:
+                    remaining_type = restrict_subtype_away(current_type, proposed_type)
+                    return {expr: proposed_type}, {expr: remaining_type}
+            else:
+                return {expr: proposed_type}, {}
+        else:
+            # An isinstance check, but we don't understand the type
+            if weak:
+                return {expr: AnyType()}, {expr: current_type}
+            else:
+                return {}, {}
+
+    def is_none(n: Node) -> bool:
+        return isinstance(n, NameExpr) and n.fullname == 'builtins.None'
+
     if isinstance(node, CallExpr):
         if refers_to_fullname(node.callee, 'builtins.isinstance'):
             expr = node.args[0]
             if expr.literal == LITERAL_TYPE:
                 vartype = type_map[expr]
                 type = get_isinstance_type(node.args[1], type_map)
-                if type:
-                    elsetype = vartype
-                    if vartype:
-                        if is_proper_subtype(vartype, type):
-                            return {expr: type}, None
-                        elif not is_overlapping_types(vartype, type):
-                            return None, {expr: elsetype}
-                        else:
-                            elsetype = restrict_subtype_away(vartype, type)
-                    return {expr: type}, {expr: elsetype}
-                else:
-                    # An isinstance check, but we don't understand the type
-                    if weak:
-                        return {expr: AnyType()}, {expr: vartype}
+                return conditional_type_map(expr, vartype, type)
+    elif (isinstance(node, ComparisonExpr) and any(is_none(n) for n in node.operands) and
+          experiments.STRICT_OPTIONAL):
+        # Check for `x is None` and `x is not None`.
+        is_not = node.operators == ['is not']
+        if is_not or node.operators == ['is']:
+            if_vars = {}  # type: Dict[Node, Type]
+            else_vars = {}  # type: Dict[Node, Type]
+            for expr in node.operands:
+                if expr.literal == LITERAL_TYPE and not is_none(expr):
+                    # This should only be true at most once: there should be
+                    # two elements in node.operands, and at least one of them
+                    # should represent a None.
+                    vartype = type_map[expr]
```

Comment:
The other crash is here; I suppose you can just do nothing if `type_map[expr]` is not present.